File: src/expression/unary.rs
use super::Expression;
use crate::token;

pub struct Unary<'a> {
    pub op: token::Token,
    pub right: &'a dyn Expression,
}

impl<'a> Unary<'a> {
    pub fn new(op: token::Token, right: &'a dyn Expression) -> Self {
        return Self { op, right };
    }
}

File: src/ast/expr/unary.go
package expr

import (
	"quick/src/error"
	"quick/src/reflect"
	"quick/src/token"
)

type Unary struct {
	Op    *token.Token
	Right Expr
}

func NewUnary(op *token.Token, right Expr) *Unary {
	return &Unary{
		Op:    op,
		Right: right,
	}
}

func (self *Unary) GetType() (reflect.Type, *error.Error) {
	return self.Right.GetType()
}

func (self *Unary) Accept(v Visitor) (*reflect.Value, *error.Error) {
	return v.VisitUnaryExpr(self)
}
